### DISCUSSION TITLE
Need help figuring out the SSO

### DISCUSSION THREAD

**Discussion by panthernet** [seraphx2/ESI.NET#58] [Troubleshooting]
Hi, currently I'm getting this error. I suppose the request should return a JSON but looks like I'm getting some binary rubbish.

```
R: ?code=ox15HMhqmEqMQk26dRyhP1&state=JrQ1R2TsXxxnPnIWrRAQ3zj6ijUpD13i

      Newtonsoft.Json.JsonReaderException: Unexpected character encountered while parsing value: . Path '', line 0, position 0.
         at Newtonsoft.Json.JsonTextReader.ParseValue()
         at Newtonsoft.Json.JsonReader.ReadForType(JsonContract contract, Boolean hasConverter)
         at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
         at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
         at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
         at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)
         at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value)
         at ESI.NET.SsoLogic.GetToken(GrantType grantType, String code, String codeChallenge) in D:\Projects\MyProjects\EVE\UFIT\ESI.NET\ESI.NET\Logic\_SSOLogic.cs:line 133
```
Here're the code scaps:

```
        [Route("auth")]
        public async Task<string> Index()
        {
             await EsiClient.SSO.GetToken(GrantType.AuthorizationCode, ControllerContext.HttpContext.Request.Query["code"]);
        }
```

![image](https://user-images.githubusercontent.com/5926603/227735502-537167b7-f4da-4568-a1da-7c1c97b59752.png)

**Comment by panthernet** [comment]
Still have that problem. Seems like every SSO related request in the lib fails. Meanwhile if I replace request code in GetToken method with this - it works.

```
using (var ssoClient = new HttpClient())
{
    ssoClient.DefaultRequestHeaders.Add("User-Agent", "BELLA");
    var values = new Dictionary<string, string> { { "grant_type", "authorization_code" }, { "code", $"{code}" } };

    ssoClient.DefaultRequestHeaders.Add("Authorization", $"Basic {_clientKey}");
    var content = new FormUrlEncodedContent(values);

    using (var tokenresponse = await ssoClient.PostAsync("https://login.eveonline.com/v2/oauth/token", content))
    {
        var responseString = await tokenresponse.Content.ReadAsStringAsync();
        var token = JsonConvert.DeserializeObject<SsoToken>(responseString);
        return token;
    }
}
```

**Comment by seraphx2** [comment]
You have no need to add the headers for user agent or authorization yourself.
First step is to setup your app according to the README file. There is an `EsiConfig` object that needs to be setup properly for this to work.

**Comment by panthernet** [comment]
> You have no need to add the headers for user agent or authorization yourself. First step is to setup your app according to the README file. There is an `EsiConfig` object that needs to be setup properly for this to work.

I have it done ofcause. Still I have the same result. I had to clone the source code and change the request code for the one shown above to make it work. It is a code from my older project. Strangely your lib doesn't work for me out of the box. 

Btw it returns code 200 and I use all the data from your config. Hmmm  except user agent maybe.

**Comment by panthernet** [comment]
I wonder what might cause this behavior? Because changing the request code makes it work.

**Comment by panthernet** [comment]
@seraphx2 
![image](https://user-images.githubusercontent.com/5926603/227740795-7226cf94-c957-4740-a374-c7014e8f26d3.png)
These two lines cause described behavior for SSO. Once I comment them all work as intended.
At least for SSO. Haven't probed other methods as I want to deal with auth first.

So I guess you need a separate HttpClient with corrected settings to work with login.eveonline.com
It's been a while since I played with EVE API so I might miss the point in my conclusion. Cheers.

**Comment by panthernet** [comment]
Nope. These two baddies also spoils corporation requests to API for me, hm.

**Comment by seraphx2** [comment]
Can you join the Discord channel to talk about this further? It is easier than going back and forth here.
